File: SalesTaxes/Model/Tax/TaxCalculator.m
//
//  TaxCalculator.m
//  SalesTaxes
//
//  Created by Fabrizio Duroni on 12/11/2016.
//  Copyright © 2016 Fabrizio Duroni. All rights reserved.
//

#import "TaxCalculator.h"
#import "TaxCalculatorContext.h"
#import "BasicTaxStrategy.h"
#import "ImportTaxStrategy.h"

#define roundUp(x) (ceil(x * 20.0f) / 20.0f)

@interface TaxCalculator ()

/// Tax Calculator Context.

@property (nonatomic, strong) TaxCalculatorContext *taxCalculatorContext;

@end

@implementation TaxCalculator

- (instancetype)init {
    
    self = [super init];
    
    if (self) {
        
        self.taxCalculatorContext = [[TaxCalculatorContext alloc] init];
    }
    
    return self;
}

- (float)getTaxes:(Product *)product {

    float taxes = 0.0f;
    
    self.taxCalculatorContext.taxCalculatorStrategy = [[BasicTaxStrategy alloc] init];
    taxes = [self.taxCalculatorContext calculateTax:product];
    
    self.taxCalculatorContext.taxCalculatorStrategy = [[ImportTaxStrategy alloc] init];
    taxes += [self.taxCalculatorContext calculateTax:product];
    
    return roundUp(taxes);
}

@end

File: SalesTaxesSwift/Model/Tax/TaxCalculator.swift
//
//  TaxCalculator.swift
//  SalesTaxes
//
//  Created by Fabrizio Duroni on 20/11/2016.
//  Copyright © 2016 Fabrizio Duroni. All rights reserved.
//

import Foundation

class TaxCalculator: TaxCalculatorDelegate {
    
    /**
     Get taxes rounded up for the product received as input.
     
     - parameter product: the product on which taxes will be calculated.
     
     - returns: taxes amount.
     */
    func getTaxes(product: Product) -> Float {
        
        var taxes:Float = 0.0
        
        var taxCalculatorContext = TaxCalculatorContext(aTaxStrategy: BasicTaxStrategy())
        taxes = taxCalculatorContext.calculateTax(product: product)
        
        taxCalculatorContext = TaxCalculatorContext(aTaxStrategy: ImportTaxStrategy())
        taxes = taxes + taxCalculatorContext.calculateTax(product: product)
        
        return self.roundUp(taxes: taxes)
    }
    
    /**
     Round up amount of taxes.
     
     - parameter taxes: taxes import to be rounded up.
     
     - returns: taxes rounded up.
     */
    private func roundUp(taxes:Float) -> Float {
        
        return (ceil(taxes * 20.0) / 20.0)
    }
}
